uberduck_ml_dev/trainer/gradtts.py

```diff
@@ -0,0 +1,247 @@
+
+# Cell
+import json
+import os
+from pathlib import Path
+from pprint import pprint
+import numpy as np
+
+import torch
+from torch.cuda.amp import autocast, GradScaler
+import torch.distributed as dist
+from torch.nn import functional as F
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.utils.data import DataLoader
+from torch.utils.tensorboard import SummaryWriter
+import time
+
+from ..models.common import MelSTFT
+from ..utils.plot import (
+    plot_attention,
+    plot_gate_outputs,
+    plot_spectrogram,
+    plot_tensor,
+)
+from ..text.util import text_to_sequence, random_utterance
+from ..text.symbols import symbols_with_ipa
+from .base import TTSTrainer
+
+from ..data_loader import (
+    TextAudioSpeakerLoader,
+    TextMelCollate,
+    DistributedBucketSampler,
+    TextMelDataset,
+)
+from ..vendor.tfcompat.hparam import HParams
+from ..utils.plot import save_figure_to_numpy, plot_spectrogram
+from ..utils.utils import slice_segments, clip_grad_value_
+from ..text.symbols import SYMBOL_SETS
+
+# Cell
+from tqdm import tqdm
+from ..text.util import text_to_sequence, random_utterance
+from ..models.gradtts import (
+    GradTTS,
+)
+from ..utils.utils import intersperse
+
+
+class GradTTSTrainer(TTSTrainer):
+    REQUIRED_HPARAMS = [
+        "training_audiopaths_and_text",
+        "test_audiopaths_and_text",
+    ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        for param in self.REQUIRED_HPARAMS:
+            if not hasattr(self, param):
+                raise Exception(f"GradTTSTrainer missing a required param: {param}")
+        self.sampling_rate = self.hparams.sampling_rate
+        self.checkpoint_path = self.hparams.log_dir
+
+    def sample_inference(self, model, timesteps=10, spk=None):
+        with torch.no_grad():
+            sequence = text_to_sequence(
+                random_utterance(),
+                self.text_cleaners,
+                1.0,
+                symbol_set=self.hparams.symbol_set,
+            )
+            if self.hparams.intersperse_text:
+                sequence = intersperse(
+                    sequence, (len(SYMBOL_SETS[self.hparams.symbol_set]))
+                )
+            x = torch.LongTensor(sequence).cuda()[None]
+            x_lengths = torch.LongTensor([x.shape[-1]]).cuda()
+            y_enc, y_dec, attn = model(
+                x,
+                x_lengths,
+                n_timesteps=50,
+                temperature=1.5,
+                stoc=False,
+                spk=spk,
+                length_scale=0.91,
+            )
+            audio = self.sample(
+                y_dec,
+                algorithm="hifigan",
+                hifigan_config="/home/w_uberduck_ai/Speech-Backbones/Grad-TTS/checkpts/hifigan-config.json",
+                hifigan_checkpoint="/home/w_uberduck_ai/Speech-Backbones/Grad-TTS/checkpts/gen_02640000_studio",
```

Comment:
pull these up to the config layer?

Reply:
Yea good idea

Reply:
Changed to support hifigan and griffin lim
```python
            if self.hparams.vocoder_algorithm == "hifigan":
                audio = self.sample(
                    y_dec,
                    algorithm=self.hparams.vocoder_algorithm,
                    hifigan_config=self.hparams.hifigan_config,
                    hifigan_checkpoint=self.hparams.hifigan_checkpoint,
                    cudnn_enabled=self.hparams.cudnn_enabled,
                )
            else:
                audio = self.sample(y_dec.cpu()[0])
            return audio
```